core/commands/tags.py

```diff
@@ -0,0 +1,318 @@
+            repo_root=self.repo_path,
+            current_head=self.get_latest_commit_msg_for_head()
+        )
+
+        cursor = len(header)
+        tags = self.get_tags()
+        regions = []
+
+        def get_region(new_text):
+            nonlocal cursor
+            start = cursor
+            cursor += len(new_text)
+            end = cursor
+            return sublime.Region(start, end)
+
+        lines = ""
+        if tags:
+            lines = "\n".join(
+                "    {} {}".format(t.sha[:7], t.tag)
+                for t in tags
+                )
+        else:
+            lines = NO_LOCAL_TAGS_MESSAGE
+
+        view_text = LOCAL_TEMPLATE.format(lines)
+        regions.append(get_region(view_text))
+
+        self.remotes = list(self.get_remotes().keys())
+        if self.remotes:
+            for remote in self.remotes:
+                remote_text = REMOTE_TEMPLATE.format(remote, NO_REMOTE_TAGS_MESSAGE)
+                regions.append(get_region(remote_text))
+                view_text += remote_text
+
+        contents = header + view_text + KEY_BINDINGS_MENU
+
+        return contents, tuple(regions)
+
+
+class GsTagsFocusEventListener(EventListener):
+
+    """
+    If the current view is a tags view, refresh the view with
+    the repository's tags when the view regains focus.
+    """
+
+    def on_activated(self, view):
+        if view.settings().get("git_savvy.tags_view") == True:
+            view.run_command("gs_tags_refresh")
+
+
+class GsTagDeleteCommand(TextCommand, GitCommand):
+
+    """
+    Delete tag(s) in selection.
+    """
+
+    def run(self, edit):
+        # Valid sections are in the Local section
+        valid_ranges = view_section_ranges[self.view.id()][:1]
+
+        lines = util.view.get_lines_from_regions(
+            self.view,
+            self.view.sel(),
+            valid_ranges=valid_ranges
+            )
+
+        items = tuple(line[4:].strip().split() for line in lines if line)
+        if items:
+            for item in items:
+                self.git("tag", "-d", item[1])
+
+            util.view.refresh_gitsavvy(self.view)
+            sublime.status_message(TAG_DELETE_MESSAGE)
+
+
+class GsTagCreateCommand(WindowCommand, GitCommand):
+
+    """
+    Through a series of panels, allow the user to add a tag and message.
+    """
+
+    def run(self):
+        sublime.set_timeout_async(self.run_async)
+
+    def run_async(self):
+        """
+        Prompt the user for a tag name.
+        """
+        self.window.show_input_panel(
+            TAG_CREATE_PROMPT,
+            "",
+            self.on_entered_tag,
+            None,
+            None
+            )
+
+    def on_entered_tag(self, tag_name):
+        """
+        After the user has entered a tag name, prompt the user for a
+        tag message. If the message is empty, use the pre-defined one.
+        """
+        # If the user pressed `esc` or otherwise cancelled.
```

Comment:
`on_entered_tag` will not be called if the user presses `esc`.  The below condition will only ever be true if the user enters an empty string.